pkg/migrate/cstor/pool.go

```diff
@@ -0,0 +1,467 @@
+	if err != nil {
+		return errors.Wrap(err, "error building kubernetes clientset")
+	}
+	c.OpenebsClientset, err = openebsclientset.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "error building openebs clientset")
+	}
+	err = c.validateCSPCOperator()
+	if err != nil {
+		return err
+	}
+	err = c.migrate(name)
+	return err
+}
+
+func (c *CSPCMigrator) validateCSPCOperator() error {
+	v1110, _ := version.NewVersion("1.11.0")
+	operatorPods, err := c.KubeClientset.CoreV1().
+		Pods(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: "openebs.io/component-name=cspc-operator",
+		})
+	if err != nil {
+		return err
+	}
+	if len(operatorPods.Items) == 0 {
+		return fmt.Errorf("cspc operator pod missing")
+	}
+	for _, pod := range operatorPods.Items {
+		operatorVersion := strings.Split(pod.Labels["openebs.io/version"], "-")[0]
+		vOperator, err := version.NewVersion(operatorVersion)
+		if err != nil {
+			return errors.Wrap(err, "failed to get operator version")
+		}
+		if vOperator.LessThan(v1110) {
+			return fmt.Errorf("cspc operator is in %s version, please upgrade it to 1.11.0 or above version",
+				pod.Labels["openebs.io/version"])
+		}
+	}
+	return nil
+}
+
+// Pool migrates the pool from SPC schema to CSPC schema
+func (c *CSPCMigrator) migrate(spcName string) error {
+	var err error
+	var migrated bool
+	c.SPCObj, migrated, err = c.getSPCWithMigrationStatus(spcName)
+	if migrated {
+		klog.Infof("spc %s is already migrated to cspc", spcName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	err = c.validateSPC()
+	if err != nil {
+		return errors.Wrapf(err, "failed to validate spc %s", spcName)
+	}
+	err = c.updateBDCLabels(spcName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to update bdc labels for spc %s", spcName)
+	}
+	klog.Infof("Creating equivalent cspc for spc %s", spcName)
+	c.CSPCObj, err = c.generateCSPC()
+	if err != nil {
+		return err
+	}
+	err = c.updateBDCOwnerRef()
+	if err != nil {
+		return err
+	}
+	// List all cspi created with reconcile off
+	cspiList, err := c.OpenebsClientset.CstorV1().
+		CStorPoolInstances(c.OpenebsNamespace).
+		List(metav1.ListOptions{
+			LabelSelector: string(apis.CStorPoolClusterCPK) + "=" + c.CSPCObj.Name,
+		})
+	if err != nil {
+		return err
+	}
+
+	// For each cspi perform the migration from csp that present was on
+	// node on which cspi came up.
+	for _, cspiItem := range cspiList.Items {
+		cspiItem := cspiItem // pin it
+		cspiObj := &cspiItem
+		err = c.cspTocspi(cspiObj)
+		if err != nil {
+			return err
+		}
+	}
+	// Clean up old SPC resources after the migration is complete
+	err = spc.NewKubeClient().
+		Delete(spcName, &metav1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// validateSPC determines that if the spc is allowed to migrate or not.
+// If the max pool count does not match the number of csp for auto spc, or
+// the bd list in spc does not match bds from the csp migration should not be done.
```

Comment:
// If the max pool count does not match the number of csp in case auto spc provisioning,
// or the blocldevice list in spc does not match bds from the csp in case of manual provisioning 
// migration can not be allowed.